### PR TITLE
Add Support for UiImage to be a Partial Texture

### LINKED ISSUE #1809: [FEATURE MODIFICATION] Add Support for UiImage to be a Partial Texture
## Feature

I would like UiImage to support the rendering of a /Sprite.

## Change

Add an extra element to the UiImage Enum, add to UI Renderable to support this, update the ui example to include a sprite.

## Reason

To reduce the number of textures needed for ui elements and to support sharing textures between sprites and ui.

## Impact

Some level of spritesheet prefab is now needed to be supported by the ui creator,

## Additional Notes

I plan to implement this issue.

I am still struggling with some of the Prefab/Creator stuff and I am currently taking any suggestions on how to implement that.

### PR DESCRIPTION
**Pull request by mrhatman**
## Description

Added the ability for UiImage to use Partial textures and sprites. 
Partial textures can be loaded via Prefab. No Prefab support for Sprite yet do to complexity in SpriteSheet Prefab.

Close #1809 

## Additions

- Partial Image to UiImage and Prefab support
- Sprite to UiImage. 

## PR Checklist

By placing an x in the boxes I certify that I have:

N/A Updated the content of the book if this PR would make the book outdated.
- [x] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Added unit tests for new code added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

If this modified or created any rs files:

- [x] Ran `cargo +stable fmt --all`
- [x] Ran `cargo clippy --all`
- [x] Ran `cargo test --all --features "empty"`

### CODE REVIEW

File: amethyst_ui/src/pass.rs [hunk 1]
```diff
@@ -525,6 +535,34 @@ fn render_image<B: Backend>(
...
+            ) {
+                batches.insert(tex_id, Some(args));
+                this_changed
+            } else {
+                false
+            }
+        }
+        UiImage::Sprite(sprite_renderer) => {
+            let sprite_sheets = resources.fetch::<AssetStorage<SpriteSheet>>();
+            let sprite_sheet = sprite_sheets.get(&sprite_renderer.sprite_sheet).unwrap();
```
Comment (Frizi) [inline amethyst_ui/src/pass.rs]: That `unwrap` is not safe. Not loaded spritesheets should simply be ignored. So this should instead be `if let Some(sprite_sheet) = ...`

File: amethyst_ui/src/image.rs [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+use amethyst_rendy::sprite::TextureCoordinates;
+use amethyst_rendy::{SpriteRender, Texture};
 
 /// Image used UI widgets, often as background.
 #[derive(Debug, Clone, PartialEq)]
 pub enum UiImage {
     /// An image backed by texture handle
     Texture(Handle<Texture>),
+    /// An image backed by partial texture
+    PartialTexture(Handle<Texture>, TextureCoordinates),
```
Comment (Frizi) [inline amethyst_ui/src/image.rs]: `An image backed by a texture cropped to specified rectangle` would be more descriptive

File: amethyst_ui/src/pass.rs (hunk 1 above)
Comment (mrhatman) [reply amethyst_ui/src/pass.rs]: I will fix that tonight.  Should I throw a warn to the logger if it isn't loaded?

Comment (Frizi) [reply amethyst_ui/src/pass.rs]: No, it can be not loaded for a while when engine works as expected. The handle can simply be empty at any time.

**Comment by azriel91** [comment]
Should be good to go if you pick up the changes from `master` (fixes CI)

**Comment by mrhatman** [comment]
bors retry

**Pull request merged**